### ISSUE TITLE
Performance problem with the security data loading

### ISSUE THREAD

**Issue by fractal3000** [jmix-framework/jmix#3994]
### Environment

Jmix version: 1.5.0

### Bug Description

Application spends a lot of time for the user role loading. Each time when we need to get the current user information application have to send a lot of requests to the database.

It seems that the process of getting a role data should be optimized. Possible the good way is to use some kind of caching.

The screens below shows that the pocess of the data loading of the user takes about 200-300 miliseconds. Such time period is a significant part of the screen opening.

![security_performance](https://github.com/user-attachments/assets/6dc58600-cf67-4582-a882-9ff1aec1cf67)

![image](https://github.com/user-attachments/assets/7c17720c-fed1-4a3d-a500-13142711f899)

io.jmix.security.impl.role.BaseRoleRepository#findRoleByCodeExcludeVisited

```
    @Nullable
    protected T findRoleByCodeExcludeVisited(String code, Set<String> visited) {
        visited.add(code);

        for (RoleProvider<T> roleProvider : getRoleProviders()) {
            T role = roleProvider.findRoleByCode(code);
            if (role != null) {
                if (role.getChildRoles() != null) {
                    for (String childCode : role.getChildRoles()) {
                        if (!visited.contains(childCode)) {
                            T childRole = findRoleByCodeExcludeVisited(childCode, visited);
                            if (childRole != null) {
                                mergeChildRoleState(role, childRole);
                            }
                        }
                    }
                }
                return role;
            }
        }
        return null;
    }
```

**Comment by glebfox** [comment]
Need to port to 2.x

**Comment by fractal3000** [comment]
This is a preposition of the problem solution
[Roles_caching.patch](https://github.com/user-attachments/files/18127130/Roles_caching.patch)

**Comment by knstvk** [comment]
For future assignee: create a separate issue in 2.x project and remove this issue from project "Release 2.6".

**Comment by alexbudarov** [comment]
UP
The performance problem is still actual, at least in Jmix 2.6.

Once you enable Hazelcast cache and you have large roles in your project - then ResourceRoleRepositoryImpl#findRoleByCode() becomes the most frequent place where UI threads spend their CPU time.

It's just called too often.

Measured in my app:
- when opening a List view of a large entity - 310 calls of ResourceRoleRepositoryImpl.findRoleByCode
- when opening a Details view of that entity - 4092 calls

**Comment by fractal3000** [comment]
## Possible solutions

### 1. Per-request role cache (RequestContextHolder)

Cache the resolved `ResourceRole` objects in the current request attributes (`RequestContextHolder`)
on first access, and reuse them for all subsequent checks within the same request.
Changes confined to one method: `AuthenticationPolicyStore.extractResourcePoliciesFromAuthenticationByScope`.

- **Effectiveness:** high — 4092 Hazelcast calls reduced to N (N = number of user's roles) per request
- **Invasiveness:** low — one method in one class
- **Architectural cleanliness:** medium — a patch on top of a flawed design, root cause not eliminated
- **Risks:** outside HTTP context (scheduled, async, `UI.access()`) `RequestContextHolder` returns null — fallback to direct repository call required; concurrent threads within a request require `ConcurrentHashMap`

---

### 2. L1 cache (Caffeine) in ResourceRoleRepositoryImpl

Add a short-lived Caffeine cache in front of Spring Cache directly in `ResourceRoleRepositoryImpl`.
Hazelcast is never reached if the role is already in JVM memory.
Equivalent to the workaround posted in the issue comments (Guava cache).

- **Effectiveness:** medium — eliminates Hazelcast network calls, but 4092 local Map lookups remain
- **Invasiveness:** low — one class; `invalidateCache()` must be extended to clear both levels
- **Architectural cleanliness:** low — two cache levels for the same data without clear justification
- **Risks:** role changes propagate with a delay equal to TTL; in a clustered deployment each node invalidates only its own L1

---

### 3. Store ResourceRole in GrantedAuthority

Resolve roles once at login time and store the compiled `ResourceRole` objects directly
in `GrantedAuthority` implementations. `AuthenticationPolicyStore` stops calling the repository
and reads the object directly from the authority.

- **Effectiveness:** very high — 0 repository calls during access checks
- **Invasiveness:** high — changes required in all authentication providers (OIDC, LDAP, SAML, REST DataStore) and in the `GrantedAuthority` creation mechanism
- **Architectural cleanliness:** high — roles are resolved exactly once at the right moment
- **Risks:** role policy changes do not apply to active sessions without re-login; a separate invalidation mechanism is needed (e.g. a global version counter checked on each request)

---

### 4. Batch API in PolicyStore

Add a method to `PolicyStore` for fetching policies for a set of attributes/entities at once.
The screen/form asks one question about all required attributes at initialization, gets one answer.
Additive change — existing methods stay, new ones are added alongside.

- **Effectiveness:** very high — one call per screen instead of thousands; roles loaded once inside the batch
- **Invasiveness:** high — new API in `PolicyStore` + migration of callers (forms, grids, constraint classes); additive, no breaking changes to existing code
- **Architectural cleanliness:** high — eliminates the root cause rather than compensating for it
- **Risks:** large scope of changes; all component initialization points where the full attribute set is known must be identified

---

### 5. Hazelcast near-cache (configuration only)

Configure near-cache for `resource-roles-cache` on the Hazelcast side — a local JVM-level copy
of frequently accessed entries. No changes to Jmix code required.

- **Effectiveness:** medium — eliminates network calls for hot entries, does not reduce call count
- **Invasiveness:** none — Hazelcast configuration in the user's application only
- **Architectural cleanliness:** neutral — a workaround specific to one cache backend
- **Risks:** Hazelcast-only; users must configure it themselves; consistency depends on near-cache TTL

---

## Combining options

**3 + 4:** the cleanest combination — roles resolved at login (option 3),
batch API (option 4) makes checks atomic and explicit. Requires the most effort
but eliminates the problem architecturally.

**1 + 4:** fix the symptom quickly with option 1, implement the batch API (option 4) gradually in parallel.
Once the batch covers all callers — option 1 can be removed.

**1 + 3:** per-request cache (option 1) relieves pressure immediately; option 3 is implemented separately
with a session invalidation mechanism. Once option 3 is in place, option 1 becomes redundant.

**2 + 1:** option 2 (Caffeine L1) eliminates Hazelcast network calls between requests,
option 1 eliminates repeated lookups within a single request. Together they give maximum effect
with minimal changes, but both remain patches on top of a flawed design.

**5 + any:** near-cache (option 5) is orthogonal to all others — it can be added on top of
any solution as an additional safeguard for Hazelcast users.

**Comment by alexbudarov** [comment]
Don't forget about hot deploy.
Role definitions can be hot-deployed in development mode, this is the main reason why they aren't cached for a long time (for the whole session lifetime).

**Comment by fractal3000** [comment]
Obviously, hot deployment is not suitable for production mode. I think it is possible to avoid unnecessary calculations or caching in this case.
